src/shogun/optimization/FirstOrderStochasticMinimizer.h

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) The Shogun Machine Learning Toolbox
+ * Written (w) 2015 Wu Lin
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ *
+ */
+#ifndef FIRSTORDERSTOCHASTICMINIMIZER_H
+#define FIRSTORDERSTOCHASTICMINIMIZER_H
+#include <shogun/optimization/FirstOrderMinimizer.h>
+#include <shogun/optimization/FirstOrderStochasticCostFunction.h>
+#include <shogun/optimization/DescendUpdater.h>
+namespace shogun
+{
+
+/** @brief The base class for stochastic first-order gradient-based minimizers.
+ *
+ * This class gives the interface of these stochastic minimizers.
+ *
+ * A stochastic minimizer is used to minimize
+ * a cost function \f$f(w)\f$ which can be written as a (finite) sum of differentiable functions, \f$f_i(w)\f$.
+ * For example,
+ * \f[
+ * f(w)=\sum_i{ f_i(w) }
+ * \f]
+ * Note that we call these differentiable functions \f$f_i(w)\f$ as sample functions.
+ *
+ * This kind of minimizers will find optimal target variables based on gradient information wrt target variables.
+ * FirstOrderStochasticMinimizer uses a sample gradient \f$\frac{\partial f_i(w) }{\partial w}\f$ to find optimal target variables,
+ * where the index \f$i\f$ is generated by some distribution.
+ *
+ * Note that FirstOrderMinimizer uses the exact gradient \f$\frac{\partial f(w) }{\partial w}\f$. 
```

Comment:
Very nice description.
Could you make sure doxygen produces links to the other classes you mention (google the syntax to reference other classes)